src/main/java/bio/terra/service/dataset/flight/create/CreateDatasetGetOrCreateBucketStep.java

```diff
@@ -0,0 +1,107 @@
+import bio.terra.stairway.StepResult;
+import bio.terra.stairway.StepStatus;
+import java.util.List;
+import java.util.Map;
+import java.util.UUID;
+import java.util.concurrent.Callable;
+import java.util.stream.Collectors;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class CreateDatasetGetOrCreateBucketStep implements Step {
+  private static final Logger logger =
+      LoggerFactory.getLogger(CreateDatasetGetOrCreateBucketStep.class);
+
+  private final AuthenticatedUserRequest userReq;
+  private final ResourceService resourceService;
+  private final DatasetRequestModel datasetRequestModel;
+  private final IamService iamService;
+
+  public CreateDatasetGetOrCreateBucketStep(
+      AuthenticatedUserRequest userReq,
+      ResourceService resourceService,
+      DatasetRequestModel datasetRequestModel,
+      IamService iamService) {
+    this.userReq = userReq;
+    this.resourceService = resourceService;
+    this.datasetRequestModel = datasetRequestModel;
+    this.iamService = iamService;
+  }
+
+  @Override
+  public StepResult doStep(FlightContext context) throws InterruptedException {
+    FlightMap workingMap = context.getWorkingMap();
+    UUID datasetId = workingMap.get(DatasetWorkingMapKeys.DATASET_ID, UUID.class);
+    GoogleRegion region = GoogleRegion.fromValueWithDefault(datasetRequestModel.getRegion());
+    UUID projectId = workingMap.get(DatasetWorkingMapKeys.PROJECT_RESOURCE_ID, UUID.class);
+    GoogleProjectResource googleProjectResource = resourceService.getProjectResource(projectId);
+
+    Callable<List<String>> getReaderGroups =
+        () ->
+            iamService
+                .retrievePolicyEmails(userReq, IamResourceType.DATASET, datasetId)
+                .entrySet()
+                .stream()
+                .filter(
+                    entry ->
+                        IngestFilePrimaryDataLocationStep.BUCKET_READER_ROLES.contains(
+                            entry.getKey()))
+                .map(Map.Entry::getValue)
+                .collect(Collectors.toList());
+
+    try {
+      GoogleBucketResource bucketForFile =
+          resourceService.getOrCreateBucketForFile(
+              region, googleProjectResource, context.getFlightId(), getReaderGroups);
+
+      workingMap.put(FileMapKeys.BUCKET_INFO, bucketForFile);
+    } catch (BucketLockException blEx) {
+      return new StepResult(StepStatus.STEP_RESULT_FAILURE_RETRY, blEx);
+    } catch (GoogleResourceNamingException ex) {
+      return new StepResult(StepStatus.STEP_RESULT_FAILURE_FATAL, ex);
+    }
+    return StepResult.getStepResultSuccess();
+  }
+
+  @Override
+  public StepResult undoStep(FlightContext context) {
+    // There is not much to undo here. It is possible that a bucket was created in the last step. We
+    // could look to
+    // see if there are no other files in the bucket and delete it here, but I think it is likely
+    // the bucket will
+    // be used again.
```

Comment:
I'm not following here -- How would the bucket be used again? Isn't the bucket still specific to the dataset created? So if the dataset create failed, then nothing would be pointing to the bucket.

Reply:
Ahh, looks like this was just a copy/paste error. In that case, is there more we can undo? (like delete the bucket?)

Reply:
Ahh, looks like this was just a copy/paste error. In that case, is there more we can undo? (like delete the bucket?)

Reply:
oh derp.  You're right.  that's a copy-paste error.  I can update.  I was wondering about this.  Actually, it looks like we don't delete the project when a dataset creation fails...ok, so let me:
a) make sure we delete the bucket here
b) add undo logic to the GetResourceBufferProject step to delete the project (that should also delete the bucket but it's cleaner to explicitly delete the bucket I think)

Does that make sense?

Reply:
Actually in the interest of not having this PR explode, I'm going to hold off on the bucket delete (we don't do that at all today...that's likely a larger story) (I'll remove the confusing undo logic).  I'll still do the project delete

Reply:
Just doing project delete makes sense to me!